SYNC-MUTATION.md

```diff
@@ -0,0 +1,335 @@
+mutable scope with a series of `set` operations.
+
+```js
+const asyncVar = new AsyncContext.Variable({ defaultValue: "default" });
+
+asyncVar.set("A"); // Throws ReferenceError: Not in a mutable context scope.
+
+// Executes the `main` function in a new mutable context scope.
+AsyncContext.contextScope(() => {
+  asyncVar.set("main");
+
+  console.log(asyncVar.get()); // => "main"
+});
+// Goes out of scope and all variables are restored in the current context.
+
+console.log(asyncVar.get()); // => "default"
+```
+
+`AsyncContext.contextScope` is basically a shortcut of
+`AsyncContext.Snapshot.run`:
+
+```js
+const asyncVar = new AsyncContext.Variable({ defaultValue: "default" });
+
+asyncVar.set("A"); // Throws ReferenceError: Not in a mutable context scope.
+
+// Executes the `main` function in a new mutable context scope.
+AsyncContext.Snapshot.wrap(() => {
+  asyncVar.set("main");
+
+  console.log(asyncVar.get()); // => "main"
+})();
+// Goes out of scope and all variables are restored in the current context.
+
+console.log(asyncVar.get()); // => "default"
+```
+
+### Use cases
+
+One use case of `set` is that it allows more intuitive test framework
+integration (or similar frameworks that have prose style declarations,
+like middlewares).
+
+```js
+describe("asynct context", () => {
+  const ctx = new AsyncContext.Variable();
+
+  beforeEach((test) => {
+    ctx.set(1);
+  });
+
+  it('run in snapshot', () => {
+    // This function is run as a second paragraph of the test sequence.
+    assert.strictEqual(ctx.get(),1);
+  });
+});
+
+function testDriver() {
+  await AsyncContext.contextScope(async () => {
+    runBeforeEach();
+    await runTest();
+    runAfterEach();
+  });
+}
+```
+
+However, without proper test framework support, mutations in async `beforeEach`
+are still unintuitive, e.g. https://github.com/xunit/xunit/issues/1880.
+
+This will need a return-value API to feedback the final context snapshot to the
+next function paragraph.
+
+```js
+describe("asynct context", () => {
+  const ctx = new AsyncContext.Variable();
+
+  beforeEach(async (test) => {
+    await undefined;
+    ctx.set(1);
+    test.setSnapshot(new AsyncContext.Snapshot());
+  });
+
+  it('run in snapshot', () => {
+    // This function is run in the snapshot saved in `test.setSnapshot`.
+    assert.strictEqual(ctx.get(),1);
+  });
+});
+
+function testDriver() {
+  let snapshot = new AsyncContext.Snapshot();
+  await AsyncContext.contextScope(async () => {
+    await runBeforeEach({
+      setSnapshot(it) {
+        snapshot = it;
+      }
+    });
+    await snapshot.run(() => runTest());
+    await runAfterEach();
+  });
+}
+```
+
+### Polyfill Viability
+
+> Can `set` be implementation in user land with `run`?
+
+The most important trait of `set` is that it will not mutate existing
+`AsyncContext.Snapshot`.
+
+A userland polyfill like the following one can not preserve this trait.
```

Comment:
Layering over unaltered AsyncContext would certainly not work, but patching snapshot retrieval can do this.